src/services/section-helper.js

```diff
@@ -0,0 +1,112 @@
+      }
+
+      // If the deletion information is exactly the same as in the ancestor, we continue the revision:
+      if (
+        section.deletedOn.getTime() === ancestorSection?.deletedOn?.getTime()
+        && section.deletedBy === ancestorSection?.deletedBy
+        && section.deletedBecause === ancestorSection?.deletedBecause
+      ) {
+        logger.info(`Section has not changed compared to ancestor section with revision ${ancestorSection.revision}, using the existing`);
+        return cloneDeep(ancestorSection);
+      }
+
+      // Otherwise we start a new revision:
+      logger.info(`Creating new revision for section key ${section.key}`);
+      return {
+        revision: uniqueId.create(),
+        key: section.key,
+        deletedOn: cloneDeep(section.deletedOn),
+        deletedBy: cloneDeep(section.deletedBy),
+        deletedBecause: section.deletedBecause,
+        type: section.type,
+        content: null
+      };
+    }
+
+    // (b) ... the continuation of a hard-deleted ancestor section
+    if (!isRestoreOperation && ancestorSection && !ancestorSection.content) {
+      return cloneDeep(ancestorSection);
+    }
+
+    // All other cases are not allowed
+    throw new Error('Sections must specify a content');
+  }
+
+  // From here on we can be sure the new section has content!
+
+  // If there is an ancestor section with the same key but the ancestor section has no content ...
+  if (ancestorSection && !ancestorSection.content) {
+
+    // ... it could be that we restore an earlier revision of a section before it has been hard-deleted
+    if (isRestoreOperation) {
+
+      // In this case we start a new revision:
+      logger.info(`Creating new revision for section key ${section.key}`);
+      return {
+        revision: uniqueId.create(),
+        key: section.key,
+        deletedOn: null,
+        deletedBy: null,
+        deletedBecause: null,
+        type: section.type,
+        content: cloneDeep(section.content)
+      };
+    }
+
+    // For all other cases we do not allow "reviving" sections
+    throw new Error(`Ancestor section with key ${section.key} is deleted and cannot be changed`);
+  }
+
+  // From here on we can be sure the new section and the ancestor section (if existing) have content!
+
+  // If there is an ancestor section with the same key and it has exactly the same content, we re-use the existing revision
+  if (ancestorSection && deepEqual(ancestorSection.content, section.content)) {
+    logger.info(`Section has not changed compared to ancestor section with revision ${ancestorSection.revision}, using the existing`);
+    return cloneDeep(ancestorSection);
+  }
+
+  // Create a new section revision:
+  logger.info(`Creating new revision for section key ${section.key}`);
+  return {
+    revision: uniqueId.create(),
+    key: section.key,
+    deletedOn: null,
+    deletedBy: null,
+    deletedBecause: null,
+    type: section.type,
+    content: cloneDeep(section.content)
+  };
```

Comment:
As this new section creation is done twice, we could extract it in a function. With this we could reduce a bit the comments (lines 76 and 101), as it will be self explanatory through the function name.

Reply:
True, I extracted all three similar cases into a function